### ISSUE TITLE
Starship process fails to spawn until a command has been run

### ISSUE THREAD

**Issue by imranmaj** [nushell/nushell#2873]
**Describe the bug**
When attempting to use a Starship prompt, the command to display the prompt in the terminal fails with the error "failed to spawn process" until a command has been run.

**To Reproduce**
1. (not sure if this is related) Install nushell 0.20.0 using `cargo install`
2. Install nushell 0.25.0 and starship 0.45.1 using `cargo install`
3. Run `config set prompt "echo $(starship prompt)"`
4. Start a new instance of nushell. There should be an error:
```
error: Failed to spawn process
  ┌─ shell:1:8
  │
1 │ echo $(starship prompt)
  │        ^^^^^^^^ failed to spawn
  ```
5. Pressing enter without entering a command will cause the error message to be printed again, as nushell attempts to display the prompt once more.

Additionally:
1. Run any command. This should cause the next prompt to display normally without error.
2. Hereafter, when pressing enter or running any subsequent commands, the prompt should display normally without error.

**Expected behavior**
The prompt should be printed normally without error when nushell starts and each time the enter key is pressed.

**Screenshots**

Nushell fails to show the prompt correctly until a command has been run:

![image](https://user-images.githubusercontent.com/49664304/103735729-86841a00-4fa3-11eb-9764-2c72ccaf695b.png)

**Configuration (please complete the following information):**

Output of `version | pivot`:

```
  #         Column0                            Column1                      
────────────────────────────────────────────────────────────────────────────
   0   version              0.25.0                                          
   1   branch               main                                            
   2   short_commit         7e4e7fa4                                        
   3   commit_hash          7e4e7fa4a610730850f050b33803a539eb3c1d57        
   4   commit_date          2021-01-05 17:32:08                             
   5   build_os             windows-x86_64                                  
   6   rust_version         rustc 1.50.0-nightly (1c389ffef 2020-11-24)     
   7   rust_channel         nightly-x86_64-pc-windows-msvc (default)        
   8   cargo_version        cargo 1.50.0-nightly (bfca1cd22 2020-11-24)     
   9   pkg_version          0.25.0                                          
  10   build_time           2021-01-05 18:13:29                             
  11   build_rust_channel   release                                         
  12   features             ctrlc, default, directories, dirs, git,         
                            ichwh, ptree, rich-benchmark, rustyline, term,  
                            uuid, which, zip                                
```

**Comment by jonathandturner** [comment]
I can repro this. Looks like the context isn't fully setup until after the first command runs

**Comment by kfardanesh** [comment]
Not a root cause, but additional info to validate the needing of running a command before starship runs.

I got this error earlier today first time after upgrading from 0.24.X. Got two errors in order as follows (paraphrased):
1. alias expected =, instead got [ ]
2. failed to spawn starship

First one was an `alias' defined in my `startup` in `config.toml`. It was simply a result of the `alias` functionality being replaced to `def`. Once I fixed that I no longer got any errors (I assume the setup of startup variables is a "first command"). For good measure, I also removed my startup variable from `config.toml` which brought back the `failed to spawn starship` error

**Comment by avandesa** [comment]
> Looks like the context isn't fully setup until after the first command runs

This might also be causing an issue I'm seeing where the current directory in the prompt will be one 'command' behind, if that makes sense.

```
~/test
❯ ls
───┬──────┬──────┬──────┬────────────
 # │ name │ type │ size │  modified
───┼──────┼──────┼──────┼────────────
 0 │ bar  │ Dir  │  0 B │ 3 secs ago
 1 │ foo  │ Dir  │  0 B │ 7 secs ago
───┴──────┴──────┴──────┴────────────

~/test
❯ cd bar

~/test
❯ pwd
C:\Users\h355305\test\bar

~/test/bar
❯ cd ..

~/test/bar
❯ pwd
C:\Users\h355305\test

~/test
❯
```

**Comment by FranzStrudel** [comment]
My current directory isn't one behind on my side

![image](https://user-images.githubusercontent.com/8066704/105970111-4affdb80-6089-11eb-995c-bc7d7ef8b1e2.png)


Note that I do have to actually enter a command (or anything really) to have starship started. If I just press enter, it repeat the error message:

![image](https://user-images.githubusercontent.com/8066704/105970320-84d0e200-6089-11eb-9efe-79679b0a3c03.png)

**Comment by mborejdo** [comment]
I am experiencing the "pwd is one behind in starship problem" aswell on Windows. I do not have the original issue from this ticket though

**Comment by tiffany352** [comment]
I've been having this issue as well.

![image](https://user-images.githubusercontent.com/1254344/106507814-6b140e00-6480-11eb-8509-5bda8c95a99f.png)

```
❯ version | pivot
────┬────────────────────┬─────────────────────────────────────────────────────────────────────────────────────────
 #  │      Column0       │                                         Column1
────┼────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────
  0 │ version            │ 0.26.0
  1 │ short_commit       │ 388973e9
  2 │ commit_hash        │ 388973e9abcdd976eb82f8739d67065011701f0e
  3 │ commit_date        │ 2021-01-26 10:07:08
  4 │ build_os           │ windows-x86_64
  5 │ rust_version       │ rustc 1.49.0 (e1884a8e3 2020-12-29)
  6 │ rust_channel       │ stable-x86_64-pc-windows-msvc (default)
  7 │ cargo_version      │ cargo 1.49.0 (d00d64df9 2020-12-05)
  8 │ pkg_version        │ 0.26.0
  9 │ build_time         │ 2021-01-26 17:38:11
 10 │ build_rust_channel │ release
 11 │ features           │ clipboard-cli, ctrlc, default, directories, dirs, ichwh, ptree, rustyline, term, trash,
    │                    │ uuid, which, zip
────┴────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────
```

```
❯ starship --version
starship 0.49.0
tag:v0.49.0
commit_hash:317226b9
build_time:2021-01-29 17:46:48
build_env:rustc 1.49.0 (e1884a8e3 2020-12-29),stable-x86_64-pc-windows-msvc (default)
```

```
❯ config get prompt
echo $(starship prompt)
```

(Nu and Starship both installed via Scoop, but I was having this issue with non-Scoop versions as well.)

**Comment by FranzStrudel** [comment]
I just fixed the issue on my side, by adding an ```startup``` entry in my config file (none was present before)

**Comment by avandesa** [comment]
I can confirm @FranzStrudel's solution works. My `startup` was failing, which was causing starship to fail. I'll open a separate ticket for the working directory being a command behind.

**Comment by fdncred** [comment]
i compiled the latest starship to try and diagnose what was going on and it's just weird. sometimes starship prompt works. sometimes it doesn't. I can't tell whether it's on our end or starships end.

I tried to be clever and create a custom command in my `startup = [ ]` section like this ` "def run_starship [] {c:\\cartar\\debug\\starship.exe prompt}",` and then change my prompt to `prompt = run_starship` and I get a starship prompt sometimes and then i change directories and the directory/pwd isn't set properly giving the impression that it's one directory behind.

**Comment by TechWatching** [comment]
I only had this issue since I upgraded nushell from 0.20 to 0.26.

**Comment by gim913** [comment]
Hi, not sure if somebody mentioned it, but I've managed to bypass first issue `will be one 'command' behind` with following hack:

`config set prompt 'with-env [PWD $(pwd)] { echo $(starship prompt) }'`

(without  additional echo, I'm getting some extra newline)

**Comment by fdncred** [comment]
I've found that this helps with extra newlines `"def starship_prompt_trim [] {starship.exe prompt | str collect | str trim}"`. Maybe some combination of both will help?

**Comment by gim913** [comment]
seems `collect` itself is enough. `trim` eats space after ">", so:
`config set prompt 'with-env [PWD $(pwd)] { starship prompt | str collect }'`

**Comment by fdncred** [comment]
Nice to have it working properly. I changed it into a custom command so I have this in my `startup = [ ]` section of the config.toml `"def starship_prompt_tweak [] {with-env [PWD $(pwd)] { starship.exe prompt | str collect }}",` and then on prompt I have `prompt = "starship_prompt_tweak"`

**Comment by jonathandturner** [comment]
@imranmaj - does this still repro with more recent versions of Nushell?

**Comment by imranmaj** [comment]
Seems to be fixed in 0.30.0

**Comment by jonathandturner** [comment]
Great, I'll go ahead and close. We can reopen if it starts happening later
